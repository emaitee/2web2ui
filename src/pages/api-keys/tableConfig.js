import React from 'react';
import { Link } from 'react-router-dom';

export const getRowData = (key) => [
  <Link to={`/account/api-keys/details/${key.id}`}>{key.label}</Link>,
  <code>{key.short_key}••••••••</code>,
  key.subaccount_id
];

export const columns = [
  { label: 'Name', width: '40%' },
  { label: 'Key', width: '20%' },
  { label: 'Subaccount', width: '20%' }
];

export const filterBoxConfig = {
  show: true,
  keyMap: { name: 'label', key: 'short_key' },
  compareKeys: ['label', 'short_key', 'id'],
  exampleModifiers: ['key', 'name', 'id']
};
