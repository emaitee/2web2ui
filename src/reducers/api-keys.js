const initialState = {
  grants: [],
  grantsLoaded: false,
  grantsLoading: false,
  keys: [],
  keysLoaded: false,
  keysLoading: false,
  error: null,
  newKey: null
};

export default (state = initialState, { payload, type }) => {
  switch (type) {
    // LIST_API_KEYS
    case 'LIST_API_KEYS_PENDING': {
      return { ...state, keysLoading: true, error: null };
    }

    case 'LIST_API_KEYS_SUCCESS': {
      return { ...state, keysLoading: false, keysLoaded: true, keys: payload };
    }

    case 'LIST_API_KEYS_FAIL': {
      return { ...state, keysLoading: false, keysLoaded: true, error: payload };
    }

    // LIST_GRANTS
    case 'LIST_GRANTS_PENDING': {
      return { ...state, grantsLoading: true, error: null };
    }

    case 'LIST_GRANTS_SUCCESS': {
      return {
        ...state,
        grants: payload,
        grantsLoaded: true,
        grantsLoading: false
      };
    }

    case 'LIST_GRANTS_FAIL': {
      return {
        ...state,
        error: payload,
        grantsLoaded: true,
        grantsLoading: false
      };
    }

    // CRUD
    case 'CREATE_API_KEY_SUCCESS':
      return { ...state, keysLoaded: false, newKey: payload.key };

    case 'DELETE_API_KEY_SUCCESS':
    case 'UPDATE_API_KEY_SUCCESS': {
      return { ...state, keysLoaded: false };
    }

    case 'HIDE_NEW_API_KEY': {
      return { ...state, newKey: null };
    }

    default: {
      return state;
    }
  }
};
