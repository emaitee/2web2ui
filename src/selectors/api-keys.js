import _ from 'lodash';
import { createSelector } from 'reselect';

/*
 * generic apiKeys selectors
 */
const getApiKeys = (state) => state.apiKeys.keys;
const getGrantsArray = (state) => state.apiKeys.grants;
const getSubaccounts = (state) => state.subaccounts.list;
const getApiKeyId = (state, props) => props.match.params.id;

const getKeysLoading = (state) => state.apiKeys.keysLoading;
const getGrantsLoading = (state) => state.apiKeys.grantsLoading;
const getSubaccountsLoading = (state) => state.subaccounts.listLoading;

export const getApiKey = createSelector(
  [getApiKeys, getApiKeyId],
  (apiKeys, id) => _.find(apiKeys, { id })
);

// Convert grants array to an object keyed by `grant.key`
export const getGrants = createSelector(getGrantsArray, (grants) =>
  _.keyBy(grants, 'key')
);

export const getLoading = createSelector(
  [getKeysLoading, getGrantsLoading, getSubaccountsLoading],
  (keysLoading, grantsLoading, subaccountsLoading) =>
    keysLoading || grantsLoading || subaccountsLoading
);

/*
 * ApiKeyForm selectors
 */
const getFormApiKey = (state, props) => props.apiKey || {};

export const getIsNew = createSelector(getFormApiKey, (apiKey) =>
  _.isEmpty(apiKey)
);

export const getInitialGrantsRadio = createSelector(
  [getGrants, getFormApiKey, getIsNew],
  (grants, apiKey, isNew) =>
    isNew || _.size(grants) <= _.size(apiKey.grants) ? 'all' : 'select'
);

export const getInitialSubaccount = createSelector(
  [getSubaccounts, getFormApiKey],
  (subaccounts, apiKey) => _.find(subaccounts, { id: apiKey.subaccount_id })
);

export const getInitialValues = createSelector(getFormApiKey, (apiKey) => {
  // using lodash methods here allow us to handle undefined values nicely.
  const grantsPairs = _.map(apiKey.grants, (grant) => [grant, 'true']);
  const grants = _.fromPairs(grantsPairs);
  const validIps = _.join(apiKey.valid_ips, ', ');

  return {
    ...apiKey,
    grants,
    validIps
  };
});
