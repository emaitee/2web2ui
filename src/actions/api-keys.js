import { formatKeyForRequest } from './helpers/api-keys';
import sparkpostApiRequest from './helpers/sparkpostApiRequest';
import { showAlert } from './globalAlert';

export function createApiKey(key) {
  return (dispatch, getState) =>
    dispatch(
      sparkpostApiRequest({
        type: 'CREATE_API_KEY',
        meta: {
          method: 'POST',
          url: '/api-keys',
          ...formatKeyForRequest(key, getState)
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key created' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not create API key',
            details: err.message
          })
        )
      );
}

export function deleteApiKey(id) {
  return (dispatch) =>
    dispatch(
      sparkpostApiRequest({
        type: 'DELETE_API_KEY',
        meta: {
          method: 'DELETE',
          url: `/api-keys/${id}`
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key deleted' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not delete API key',
            details: err.message
          })
        )
      );
}

export function updateApiKey(id, key) {
  return (dispatch, getState) =>
    dispatch(
      sparkpostApiRequest({
        type: 'UPDATE_API_KEY',
        meta: {
          method: 'PUT',
          url: `/api-keys/${id}`,
          ...formatKeyForRequest(key, getState)
        }
      })
    )
      .then(() =>
        dispatch(showAlert({ type: 'success', message: 'API key updated' }))
      )
      .catch((err) =>
        dispatch(
          showAlert({
            type: 'error',
            message: 'Could not update API key',
            details: err.message
          })
        )
      );
}

export function hideNewApiKey() {
  return {
    type: 'HIDE_NEW_API_KEY'
  };
}

export function listApiKeys({ force } = {}) {
  return (dispatch, getState) => {
    if (!force && getState().apiKeys.keysLoaded) {
      return;
    }

    return dispatch(
      sparkpostApiRequest({
        type: 'LIST_API_KEYS',
        meta: {
          method: 'GET',
          url: '/api-keys'
        }
      })
    );
  };
}

export function listGrants({ force } = {}) {
  return (dispatch, getState) => {
    if (!force && getState().apiKeys.grantsLoaded) {
      return;
    }

    return dispatch(
      sparkpostApiRequest({
        type: 'LIST_GRANTS',
        meta: {
          method: 'GET',
          url: '/authenticate/grants'
        }
      })
    );
  };
}
