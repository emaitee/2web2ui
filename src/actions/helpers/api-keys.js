import _ from 'lodash';
import { getGrants } from 'src/selectors/api-keys';

export function formatKeyForRequest(key, getState) {
  const request = { data: {}};

  if (key.subaccount) {
    request.headers = { 'X-MSYS-SUBACCOUNT': key.subaccount.id };
  }

  request.data.label = key.label;

  if (key.grantsRadio === 'all') {
    request.data.grants = _.keys(getGrants(getState()));
  } else {
    request.data.grants = _.keys(_.pickBy(key.grants));
  }

  if (key.validIps) {
    request.data.valid_ips = key.validIps.split(',').map(_.trim);
  }

  return request;
}
