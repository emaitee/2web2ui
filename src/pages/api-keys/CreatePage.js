import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { createApiKey } from 'src/actions/api-keys';
import Layout from 'src/components/layout/Layout';
import { getLoading } from 'src/selectors/api-keys';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/api-keys'
};

export class CreatePage extends React.Component {
  onSubmit = (values) => {
    const { createApiKey, history } = this.props;

    return createApiKey(values).then((res) => {
      history.push('/account/api-keys');
    });
  };

  render() {
    return (
      <Layout.App loading={this.props.loading}>
        <Page title="Create API Key" breadcrumbAction={breadcrumbAction} />
        <Panel>
          <Panel.Section>
            <ApiKeyForm onSubmit={this.onSubmit} />
          </Panel.Section>
        </Panel>
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => ({
  loading: getLoading(state)
});

export default withRouter(
  connect(mapStateToProps, { createApiKey })(CreatePage)
);
