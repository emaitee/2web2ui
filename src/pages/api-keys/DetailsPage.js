import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Page, Panel } from '@sparkpost/matchbox';

import { deleteApiKey, listApiKeys, updateApiKey } from 'src/actions/api-keys';
import Layout from 'src/components/layout/Layout';
import DeleteModal from 'src/components/deleteModal/DeleteModal';
import { getApiKey, getLoading } from 'src/selectors/api-keys';
import ApiKeyForm from './components/ApiKeyForm';

const breadcrumbAction = {
  content: 'API Keys',
  Component: Link,
  to: '/account/api-keys'
};

export class ApiKeysDetailsPage extends Component {
  static defaultProps = {
    apiKey: {}
  };

  state = {
    showDeleteModal: false
  };

  constructor(props) {
    super(props);
    this.secondaryActions = [
      { content: 'Delete', onClick: this.onToggleDelete }
    ];
  }

  componentDidMount() {
    this.props.listApiKeys();
  }

  onDelete = () => {
    const { deleteApiKey, history } = this.props;

    deleteApiKey().then(() => {
      history.push('/account/api-keys');
    });
  };

  onToggleDelete = () => {
    this.setState({ showDeleteModal: !this.state.showDeleteModal });
  };

  onSubmit = (values) => {
    const { updateApiKey, history } = this.props;

    return updateApiKey(values).then((res) => {
      history.push('/account/api-keys');
    });
  };

  render() {
    const { apiKey, loading } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page
          title={apiKey.label}
          breadcrumbAction={breadcrumbAction}
          secondaryActions={this.secondaryActions}
        />
        <Panel>
          <Panel.Section>
            <ApiKeyForm apiKey={apiKey} onSubmit={this.onSubmit} />
          </Panel.Section>
        </Panel>
        <DeleteModal
          open={this.state.showDeleteModal}
          title="Delete API Key"
          text="Are you sure you want to delete this API Key?"
          handleToggle={this.onToggleDelete}
          handleDelete={this.onDelete}
        />
      </Layout.App>
    );
  }
}

const mapStateToProps = (state, props) => {
  const { error, grants, keys } = state.apiKeys;

  return {
    apiKey: getApiKey(state, props),
    keys,
    error,
    grants,
    loading: getLoading(state)
  };
};

const mapDispatchToProps = (dispatch, props) => {
  const { id } = props.match.params;

  return {
    deleteApiKey: () => dispatch(deleteApiKey(id)),
    listApiKeys: () => dispatch(listApiKeys()),
    updateApiKey: (values) => dispatch(updateApiKey(id, values))
  };
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ApiKeysDetailsPage)
);
