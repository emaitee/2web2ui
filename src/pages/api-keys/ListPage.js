import copy from 'copy-to-clipboard';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Banner, Page } from '@sparkpost/matchbox';

import { hideNewApiKey, listApiKeys } from 'src/actions/api-keys';

import ApiErrorBanner from 'src/components/apiErrorBanner/ApiErrorBanner';
import TableCollection from 'src/components/collection/TableCollection';
import Layout from 'src/components/layout/Layout';
import { getLoading } from 'src/selectors/api-keys';
import PermissionsColumn from './components/PermissionsColumn';

const columns = ['Name', 'Key', 'Permissions'];

const primaryAction = {
  content: 'Create API Key',
  Component: Link,
  to: '/account/api-keys/create'
};

const getRowData = (key) => [
  <Link to={`/account/api-keys/details/${key.id}`}>{key.label}</Link>,
  <code>{key.short_key} ••••</code>,
  <PermissionsColumn grants={key.grants} />
];

export class ListPage extends Component {
  state = { copied: false };

  componentDidMount() {
    this.props.listApiKeys();
  }

  onClickBanner = () => {
    copy(this.props.newKey);
    this.setState({ copied: true });
  };

  onReloadApiBanner = () => {
    this.props.listApiKeys({ force: true }); // force a refresh
  };

  renderBanner() {
    const { hideNewApiKey, newKey } = this.props;

    const action = {
      content: this.state.copied ? 'Copied to clipboard' : 'Copy',
      onClick: this.onClickBanner
    };

    return (
      <Banner
        action={action}
        title="New API Key"
        status="success"
        onDismiss={hideNewApiKey}
      >
        <p>
          Make sure to copy your API key now. You won't be able to see it again!
        </p>
        <strong>{newKey}</strong>
      </Banner>
    );
  }

  renderCollection() {
    const { keys } = this.props;

    return (
      <TableCollection
        columns={columns}
        getRowData={getRowData}
        pagination={true}
        rows={keys}
      />
    );
  }

  renderError() {
    return (
      <ApiErrorBanner
        errorDetails={this.props.error.message}
        message="Sorry, we seem to have had some trouble loading your API keys."
        reload={this.onReloadApiBanner}
      />
    );
  }

  render() {
    const { error, loading, newKey } = this.props;

    return (
      <Layout.App loading={loading}>
        <Page primaryAction={primaryAction} title="API Keys" />
        {newKey && this.renderBanner()}
        {error ? this.renderError() : this.renderCollection()}
      </Layout.App>
    );
  }
}

const mapStateToProps = (state) => {
  const { error, keys, newKey } = state.apiKeys;
  return {
    error,
    keys,
    loading: getLoading(state),
    newKey
  };
};

export default connect(mapStateToProps, { hideNewApiKey, listApiKeys })(
  ListPage
);
