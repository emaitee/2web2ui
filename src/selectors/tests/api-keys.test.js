import { getFilteredKeys } from '../api-keys';

describe('Filter Api Keys Selector', () => {
  const store = {
    apiKeys: {
      keys: [
        {
          label: 'One',
          subaccount_id: 1
        },
        {
          label: 'Two',
        },
      ]
    },
    form: {
      apiKeysFilters: {}
    }
  };

  it('returns unfiltered list', () => {
    expect(getFilteredKeys(store)).toMatchSnapshot();
  });

  it('filters search', () => {
    const search = { form: { apiKeysFilters: { values: { search: 'tw' }}}};
    expect(getFilteredKeys({...store, ...search})).toMatchSnapshot();
  });

  it('filters assigned to subaccount', () => {
    const search = { form: { apiKeysFilters: { values: { subaccount: { subaccount: true} }}}};
    expect(getFilteredKeys({...store, ...search})).toMatchSnapshot();
  });

  it('filters assigned to master', () => {
    const search = { form: { apiKeysFilters: { values: { subaccount: { master: true} }}}};
    expect(getFilteredKeys({...store, ...search})).toMatchSnapshot();
  });

  it('filters all subaccount options', () => {
    const search = { form: { apiKeysFilters: { values: { subaccount: { master: true, subaccount: true }}}}};
    expect(getFilteredKeys({...store, ...search})).toMatchSnapshot();
  });
});
