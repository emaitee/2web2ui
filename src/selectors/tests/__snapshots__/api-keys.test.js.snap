// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Api Keys Selector filters all subaccount options 1`] = `
Array [
  Object {
    "label": "One",
    "subaccount_id": 1,
  },
  Object {
    "label": "Two",
  },
]
`;

exports[`Filter Api Keys Selector filters assigned to master 1`] = `
Array [
  Object {
    "label": "Two",
  },
]
`;

exports[`Filter Api Keys Selector filters assigned to subaccount 1`] = `
Array [
  Object {
    "label": "One",
    "subaccount_id": 1,
  },
]
`;

exports[`Filter Api Keys Selector filters search 1`] = `
Array [
  Object {
    "label": "Two",
  },
]
`;

exports[`Filter Api Keys Selector returns unfiltered list 1`] = `
Array [
  Object {
    "label": "One",
    "subaccount_id": 1,
  },
  Object {
    "label": "Two",
  },
]
`;
